cli/cli_app.go

```diff
@@ -670,6 +674,17 @@ func validateAll(terragruntOptions *options.TerragruntOptions) error {
 	return stack.Validate(terragruntOptions)
 }
 
+// checkProtectedModule checks if module is protected via the "prevent_destroy" flag
+func checkProtectedModule(terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig) error {
+	if firstArg(terragruntOptions.TerraformCliArgs) != "destroy" {
+		return nil
+	}
+	if terragruntConfig.PreventDestroy {
+		return errors.WithStackTrace(ModuleIsProtected{Opts: terragruntOptions})
```

Comment:
Hm, there's a bit of gotcha here, when someone runs `destroy-all`, this module will exit with an error, causing anything that depends on it to exit with an error, and causing the overall process to exit with an error. On the one hand, this is good, for if module B depends on A, and you put `prevent_destroy` in B, then not only does that protect B, but it'll also prevent A from being destroyed (without which B would also break!). On the other hand, if this error prevents unrelated parts of the dependency graph from executing, that would be unfortunate.

Could you add a test for this? E.g., add a test fixture with:

- Module A
- Module B which depends on A and has `prevent_destroy`
- Module C
- Module D which depends on C
- Module E

If you run `destroy-all` on this graph, my hope is everything but module A and B would be deleted...

Reply:
@brikis98 it seems that I stuck there.

How is better to check if module was destroyed in the integration tests?
I created separate directory with fixtures of modules (that are using `hello-world` source), I can easily run `destroy-all` on them and check if the error has kind of `configstack.MultiError`.

But how can I check if module C, D, E were actually destroyed or not?

Reply:
I can think of a few options. One way to do it is to run `terragrunt output` or `terragrunt show` in each module, which should tell you if it has been applied or not. After running `apply-all`, you should see all have been applied; after running `destroy-all`, you should see that C, D, and E have been destroyed.

Reply:
@brikis98 Thank you for the advice with `terragrunt show` (I just started working with Terragrunt, so I haven't got all the stuff yet).
I added the `TestPreventDestroyDependencies` that tests several modules.